### PR TITLE
feat(logging): Add simple support for CoAP request logging

### PR DESCRIPTION
**Pull request by akolosov-n**
Add RequestLoggerFilter to provide request logging with customizable log message format.

### CODE REVIEW

File: coap-core/src/main/java/com/mbed/coap/server/CoapServerBuilder.java [hunk 1]
```diff
@@ -120,7 +120,7 @@ public CoapServerBuilder routeFilter(Filter.SimpleFilter<CoapRequest, CoapRespon
         return this;
     }
 
-    public CoapServerBuilder inboundRequestFilter(Filter.SimpleFilter<CoapRequest, CoapResponse> inboundRequestFilter) {
+    public CoapServerBuilder inboundRequestFilter(Filter<CoapRequest, CoapResponse, CoapRequest, CoapResponse> inboundRequestFilter) {
```
Comment (szysas) [inline coap-core/src/main/java/com/mbed/coap/server/CoapServerBuilder.java:123]: What is the reason for this change?

File: coap-core/src/main/java/com/mbed/coap/server/filter/RequestLoggerFilter.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+    @FunctionalInterface
+    public interface Formatter {
+        String apply(CoapRequest req, CoapResponse resp, long duration);
+    }
+
+    public static class Builder {
+        private Logger logger = LoggerFactory.getLogger(RequestLoggerFilter.class);
+        private Level logLevel = Level.INFO;
+        private Formatter msgFormatter = (req, resp, duration) -> String.format("%s -> %s (%dms)", req, resp, duration);
```
Comment (szysas) [inline coap-core/src/main/java/com/mbed/coap/server/filter/RequestLoggerFilter.java:55]: Could this be extracted to static function with some 'default' name. 
And then it becomes testable ;)
Also could we have IP address at the beginning of a log message

File: coap-core/src/main/java/com/mbed/coap/server/CoapServerBuilder.java (hunk 1 above)
Comment (akolosov-n) [reply coap-core/src/main/java/com/mbed/coap/server/CoapServerBuilder.java:123]: Without this I couldn't pass `filter.andThen(otherFilter)` to `CoapServerBuilder.inboundRequestFilter()`.

Comment (akolosov-n) [reply coap-core/src/main/java/com/mbed/coap/server/CoapServerBuilder.java:123]: Well, this can be worked around by just casting to `Filter.SimpleFilter` of the `andThen()` result.

**Review by szysas — APPROVED** [review]
